Python/chapter01/1.7 - Rotate Matrix/miguel_1.7_sol.py

```diff
@@ -0,0 +1,172 @@
+"""
+Python version 3.7.0
+1.7 - Rotate Matrix
+Given an image represented by an NxN matrix, where each pixel in the image is 4 bytes, write a method to rotate
+the image by 90 degrees.  Can you do this in place?
+"""
+import unittest
+from typing import List
+
+
+def rotate_matrix(matrix: List[List[int]]) -> List[List[int]]:
+    """
+    Rotate matrix will rotate the given matrix by 90 degrees.
+    Runtime: O(N^2), asymptotic runtime depends on N
+    Space Complexity: O(N^2), creating a new matrix of NxN called 'rotated'
+    :param matrix: an NxN matrix
+    :param N: the size of the matrix (NxN)
+    :return: a newly rotated matrix
+    """
+    N = len(matrix)
+    rotated = [[0 for i in range(N)] for j in range(N)]
+    for i in range(N):
+        for j in range(N):
+            rotated[j][(N-1)-i] = matrix[i][j]
+    return rotated
+
+
+def _perform_full_rotation(matrix: List[List[int]], row: int, col: int, M: int):
+    """
+    Helper function that performs four 90 degree rotations starting at row, col.
+    :param matrix: an MxM sub-matrix out of an NxN matrix
+    :param row: starting row
+    :param col: starting column
+    :param M: size of sub-matrix
+    :return:
+    """
+    num_rotations = 4
+    start_row = row
+    rotated_row = row
+    rotated_col = col
+    temp_new = matrix[row][col]
+    for r in range(0, num_rotations):
+        # compute new rotated indices
+        prev_col = rotated_col
+        rotated_col = M - 1 - rotated_row + (start_row * 2)  # offset to account for reduced N
+        rotated_row = prev_col
+        # store value at newly computed indices
+        temp_new, matrix[rotated_row][rotated_col] = matrix[rotated_row][rotated_col], temp_new
+
+
+def rotate_matrix_in_place(matrix: List[List[int]]) -> List[List[int]]:
+    """
+    Does the same as rotate_matrix, but in place.
+    Runtime: O(N^2), asymptotic runtime depends on N. We make N^2 swaps.
+    Space Complexity: O(1), constant amount of temp variables that does not depend on N.
+    :param matrix: an NxN matrix
+    :return: the input matrix, but rotated
+    """
+    start_row = 0
+    N = len(matrix)
+    for n in range(N, 1, -2):
+        for col in range(start_row, n - 1 + start_row):
+            _perform_full_rotation(matrix, start_row, col, n)
```

Comment:
you might want to add some comments that overall explain how this method works since it's non-trivial and its workings are split across methods. this increases local readability at the expense of global understandability, but that can be offset with comments so it's the right trade-off to do.

you can mention that every iteration of the outer loop will rotate one of the "rings" of the matrix, the inner loop iterates rotating 4 entries of the ring every time, and you can describe how you rotate the 4 entries either here or in a comment in `_perform_full_rotation`. bonus points for some neat [ASCII diagram](https://blog.regehr.org/archives/1653), but you don't need to go that far if you can convey the idea in prose.

you can see an example of how comments can be used to explain a gnarly piece of code (plus an instance of me judiciously breaking the rule of "prefer loops over recursion"!) here: https://android.googlesource.com/platform/external/minijail/+/255a8e2879edb49f395f819e2f8590b40a8284ed/tools/compiler.py#158

Reply:
Will add comments, thank you for all the feedback!

Reply:
Added comments, sorry for the delay, I had exams and many assignments to complete last week.